### ISSUE TITLE
[selectors] `:lang` for documents without content language and for elements of unknown language; consider `:lang("")` over `:not(:lang("*"))`

### ISSUE THREAD

**Issue by myfonj** [w3c/csswg-drafts#6915]
### Brief questions to answer

1. Is `:lang("*")` really valid selector?  (Safari supports it and Chrome accepted [Issue 1281157](https://bugs.chromium.org/p/chromium/issues/detail?id=1281157) to implement it.)
2. How to address document that failed to define it's content language?
3. Should be HTML `<el lang=""></el>` or equivalent `<el lang></el>` matched by CSS `:lang("")` or even `:lang()`?  (Not yet proposed nor implemented.)
4. Is erroneous "undefined" document language equivalent of `<html lang="">` or is it something different?

### Trivia

1. It seems that there have never been a way for `lang()` functional pseudo-class to precisely target document without defined content language nor element sub-tree set as such with `lang=""` attribute.
2. [Selectors-4 draft introduces wildcard support](https://drafts.csswg.org/selectors-4/#the-lang-pseudo) ([md](https://github.com/w3c/csswg-drafts/blob/da9f4c11ca40e850c22dd8387e4069cb818e702e/selectors-4/Overview.bs#L1896)) in string argument for matching "any" language of given script or region group (like `"*-Latn"` or `"*-ch"`), opening possibility (loophole?) for matching *any **specified** language* value using `:lang("*")` and with conjunction with negation pseudo-class opens possibility to use `:not(:lang("*"))` for targeting elements that belong to "no specified language" from previous point.  While plain `"*"` value is not explicitly mentioned in the draft, this reportedly already works in current Safari. 
3. Undetermined language occur on any HTML document that does not have explicit `lang` attribute and does not come with `content-language` HTTP header or it's "deprecated nasty" `<meta http-equiv>` counterpart.
4. For marking ~~non-linguistic~~ __unknown language__ content it is advised to use `lang=""` attribute. While it is possible to target it with attribute selector (`[lang=""]`), is does not seem like a right tool and introduces nesting / inheritance problems that `lang()` selector was created for.

#### Sample 1: Document without content language

```
HTTP/2 200 OK
[no `content-language: xy,zz` HTTP header here]

<html [no `lang="xy"]` here>
	<head>
		[no `<meta http-equiv="content-language" content="xy,zz">` here]
	</head>
	<body>
		[I want target this document.]
	</body>
</html>
```

#### Sample 2: element with unknown language content

```HTML
<html lang="xy">
<body>
 <p>xyx:
  <samp lang="">
   000<em>111<var lang="xy">x</var></em>000
  </samp>
 </p>
 [I want to target elements with digits and omit elements with letters.]
</body>
</html>
```

### Problems

Inability to "legally" target specifically undetermined language document or element seems to be quite minor issue, since most CSS approaches tend to start with "common" defaults and progress to language specific with selector of higher specificity.  Like

#### Sample 3: using specificity or order for refining defaults

```CSS
/* pseudo example of common approach for setting language related styles */
/* hand-waving default */ :root { quotes: '"' '"' "'" "'"; } 
/* specific known language */ :lang(x-whatever) { quotes "→" "←" "☛" "☚"; }
```

However this approach fails short if CSS' author needs to specifically address **poorly marked-up document** lacking any content language hint

#### Sample 4: using `body:not(:lang("*"))` for styling "bad" document

```CSS
/* Pseudo example "let's put a country flag representing language of the document on it's beginning" */
/* known languages */
body:lang(en-gb)::before { content: url(./flags-gb.svg) / "British English document: "; }
body:lang(en-us)::before { content: url(./flags-us.svg) / "American English document: "; }
/* [etc] */

/* default for "unknown" */
body::before { content: url(./flags-specified-but-unknown.svg) / "Content language of this document is specified but not included in style sheets - please contact style maintainer. "; font-size: small; color: GrayText; }

/* "error message" for "unspecified". topic of this issue */
body:not(:lang("*"))::before { content: url(./flags-unspecified-error.svg) / "This document appears to have no content language specified. If you are it's maintainer fix it ASAP, please. "; }`
body:not(:lang("*")) { background-color: var(--bg-error, Canvas); color: var(--fg-error, GrayText); font-family: cursive; }
```

(Think Sample 3 above. Also notice that samples use CSS features not yet widely implemented - pseudo element alternative texts and system colors - but those are not related for the issue in question.)

### Links, resources and notes

- my initial [SO question](https://stackoverflow.com/questions/70354616/css-lang-selector-for-elements-in-documents-of-undetermined-language) - thanks onkar ruikar for answer, analysis, discussion, workaround and chromium bugreport
- [tweet](https://twitter.com/myfonj/status/1472362381338628099)
- https://www.w3.org/International/questions/qa-css-lang
- https://www.w3.org/International/questions/qa-no-language
  - "unknown": `lang=""` / `xml:lang="und"`
  - "known but non-linguistic": `lang="zxx"`; 
- https://www.w3.org/International/articles/language-tags/
- [rfc2070: Internationalization of the Hypertext Markup Language](https://www.ietf.org/rfc/rfc2070.txt)

**Comment by frivoal** [comment]
This was discussed at TPAC by the i18n-WG, and the conclusion was that we should have:
* `:lang("")` that matches `lang=""` (and descendants…)
* `:lang("*")` that matches everything but `lang=""` and descendants
* a note about `lang="und"` and `lang=""` being treated distinctly, despite having similar semantics

**Comment by dbaron** [comment]
`und` is discussed [in RFC 5646](https://www.rfc-editor.org/rfc/rfc5646.html#page-56)

**Comment by fantasai** [comment]
Fixed in https://github.com/w3c/csswg-drafts/commit/9b51686163492af04c1954aaf668d5d24cc2d165

**Comment by aphillips** [comment]
@fantasai @frivoal Thanks for this fix. Alas, I just now posted some comments related to the change. I can file a new issue if needed.

@frivoal I'm closing [your action](https://www.w3.org/International/track/actions/1197) for I18N...

**Comment by myfonj** [comment]
Should we reopen this issue? Main changes made in 

> Fixed in [9b51686](https://github.com/w3c/csswg-drafts/commit/9b51686163492af04c1954aaf668d5d24cc2d165)

were overwritten by [late merge of #4212](https://github.com/w3c/csswg-drafts/pull/4212/files#diff-ff7e41817bff012f3ddff1f611038e554675a63acfb0ae367398437958b0baf4L1987-R1997), so the explicit distinction of `lang("*")` vs `lang("")` is quite gone now.

**Comment by tabatkins** [comment]
Ah, thanks for the pointer. I went ahead and restored that text, it was definitely an accidental removal.
